### ISSUE TITLE
[BUG]  The energy values obtained from the dp test command do not match the energy values obtained from dp's Python interface dp.eval.

### ISSUE THREAD

**Issue by Zch102xjtumse** [deepmodeling/deepmd-kit#4733]
### Bug summary

The energy values obtained from the dp test command do not match the energy values obtained from dp's Python interface dp.eval.
I referred to discussions #3903 and #3020 regarding consistency between data types and model types. After reading the data(the code as following)
import torch
import dpdata
from deepmd.pt.infer.deep_eval import DeepPot as DP
from deepmd.infer import calc_model_devi
from deepmd.tf.infer.model_devi import write_model_devi_out
import numpy as np
import os
import re
import sys
import datetime
import glob
        data=dpdata.LabeledSystem(path,fmt="deepmd/npy")
        real_type = np.loadtxt(path+"/type_map.raw", dtype=str)
        cell=data['cells']
        coords=data['coords']
        atype_from_data = data['atom_types']
        e0=data['energies']
        f0=data['forces']
 I rearranged the data order according to the model's type sequence, but I still obtained different results.

### DeePMD-kit Version

DeePMD-kit v3.0.0

### Backend and its version

Pytorch 2.4.1

### How did you download the software?

Offline packages

### Input Files, Running Commands, Error Log, etc.

The result of dp.eval(focus on model1):
![Image](https://github.com/user-attachments/assets/c9e4dec8-6614-46d1-81d1-c3ba579e726b)
The result of dp test(model1):
![Image](https://github.com/user-attachments/assets/3659a3e8-475c-496c-9a2d-6204f0101e31)

### Steps to Reproduce

import torch
import dpdata
#from deepmd.infer import DeepPot as DP
from deepmd.pt.infer.deep_eval import DeepPot as DP
from deepmd.infer import calc_model_devi
from deepmd.tf.infer.model_devi import write_model_devi_out
import numpy as np
import os
import re
import sys
import datetime
import glob

def predict_DP(coords,cells,atype,i):
    model_type= DP.get_type_map(GRAPH[i])
    print("model:",i,"model_type:",model_type)
    e, f, v = GRAPH[i].eval(coords, cells, atype)
    return e, f

def calc_std(de_list, df_list):
    de_std = np.std(de_list, axis=0)
    df_std = np.std(df_list, axis=0)
    return de_std, df_std

# 加载模型
model_path = '/home/zhaochenhao/job/DP-potential/CrCoNi/v9/data-0.1/iteration-2/'
model_files = sorted(glob.glob(os.path.join(model_path, '*.pth')))

for model_file in model_files:
    print(f"- {model_file}")

GRAPH = []
for model_file in model_files:
    GRAPH.append(DP(model_file))
    
num_models = len(GRAPH)

# 数据文件夹路径
folder_path = '/home/zhaochenhao/job/DP-potential/CrCoNi/v9/data-0.1/check-code'
current_dir = os.path.dirname(os.path.abspath(__file__))

# 输出文件
timestamp = datetime.datetime.now().strftime("%Y%m%d_%H%M%S")
energy_diff_file = os.path.join(current_dir, f"gsfe-test-energy_diff_{timestamp}.txt")
force_diff_file = os.path.join(current_dir, f"gsfe-test-force_diff_{timestamp}.txt")
model_devi_file = os.path.join(current_dir, f"gsfe-test-model_devi_{timestamp}.txt")
energy_values_file = os.path.join(current_dir, f"gsfe-test-energy_values_{timestamp}.txt")
force_values_file = os.path.join(current_dir, f"gsfe-test-force_values_{timestamp}.txt")

# 创建能量差值文件
with open(energy_diff_file, "w") as f:
    f.write("# 数据集路径\n")
    f.write("#       frame")
    for m in range(num_models):
        f.write(f"         model{m+1}_devi_e")
    f.write("\n")

# 创建力差值文件
with open(force_diff_file, "w") as f:
    f.write("# 数据集路径\n")
    f.write("#       frame       atom_id")
    for m in range(num_models):
        f.write(f"         model{m+1}_fx         model{m+1}_fy         model{m+1}_fz")
    f.write("\n")

# 创建每个数据点的模型标准差文件
with open(model_devi_file, "w") as f:
    f.write("# 数据集路径\n")
    f.write("#       frame         devi_e         devi_fx         devi_fy         devi_fz\n")

# 创建能量预测值和真值文件
with open(energy_values_file, "w") as f:
    f.write("# 数据集路径\n")
    f.write("#       frame         true_e")
    for m in range(num_models):
        f.write(f"         model{m+1}_e")
    f.write("\n")

# 创建力预测值和真值文件
with open(force_values_file, "w") as f:
    f.write("# 数据集路径\n")
    f.write("#       frame       atom_id         true_fx         true_fy         true_fz")
    for m in range(num_models):
        f.write(f"         model{m+1}_fx         model{m+1}_fy         model{m+1}_fz")
    f.write("\n")

for i in os.listdir(folder_path):
    p1=os.path.join(folder_path,i)
    for j in os.listdir(p1):
        path=os.path.join(p1,j)
        print(path)
        data=dpdata.LabeledSystem(path,fmt="deepmd/npy")
        real_type = np.loadtxt(path+"/type_map.raw", dtype=str)
        # print("real_type:",real_type)
        cell=data['cells']
        coords=data['coords']
        # print("Original coords shape:", coords.shape)
        atype_from_data = data['atom_types']
        e0=data['energies']
        f0=data['forces']
        
        target_element_order = ['Cr', 'Co', 'Ni', 'H']
        
        current_actual_atom_symbols = [real_type[t] for t in atype_from_data]
        # print(current_actual_atom_symbols)
        atoms_to_sort_info = []
        for original_idx, symbol_in_data in enumerate(current_actual_atom_symbols):
            try:
                priority = target_element_order.index(symbol_in_data)
                # print("priority",priority)
                atoms_to_sort_info.append({'symbol': symbol_in_data, 
                                           'original_index': original_idx, 
                                           'priority': priority})
            except ValueError:
                error_message = (f"Atom symbol '{symbol_in_data}' (from type_map.raw and atom_types) "
                                 f"at original index {original_idx} is not defined in target_element_order: "
                                 f"{target_element_order}. Reordering and re-typing cannot proceed.")
                print(error_message)
                raise ValueError(error_message)

        sorted_atom_info_list = sorted(atoms_to_sort_info, key=lambda x: x['priority'])

        permutation_indices = [info['original_index'] for info in sorted_atom_info_list]

        coords = coords[:, permutation_indices, :] 
        f0 = f0[:, permutation_indices, :]     
        
        final_symbols_in_new_order = [info['symbol'] for info in sorted_atom_info_list]
        

        atype = np.array([target_element_order.index(s) for s in final_symbols_in_new_order])
        
        # 获取原子数量
        num_atoms = len(atype)
        print(f"构型原子数 (after reordering): {num_atoms}")
        print(f"New atype for DP (0:Cr, 1:Co, 2:Ni, 3:H if present): {atype}")

        de_list = []
        df_list = []
        
        for m in range(num_models):
            e1, f1 = predict_DP(coords, cell, atype, m)
            de = (e1 - e0) / num_atoms
            df = f1 - f0
            de_list.append(de)
            df_list.append(df)

            
        with open(energy_values_file, "a") as f:
            # 添加当前数据集路径的标识行
            f.write(f"# {path}\n")
            for frame_idx in range(len(e0)):
                # 获取真实能量值（除以原子数）
                true_e = float(e0[frame_idx]) if np.isscalar(e0[frame_idx]) else float(np.mean(e0[frame_idx]))
                true_e_per_atom = true_e / num_atoms
                
                # 获取各模型预测的能量值（除以原子数）
                model_e_values = []
                for m in range(num_models):
                    # 计算预测值 = 差值 + 真值
                    pred_e = de_list[m][frame_idx] + e0[frame_idx]
                    e_val = float(pred_e) if np.isscalar(pred_e) else float(np.mean(pred_e))
                    e_val_per_atom = e_val / num_atoms
                    model_e_values.append(e_val_per_atom)
                
                # 写入文件
                f.write(f"{frame_idx:12d}    {true_e_per_atom:15.6e}")
                for e_val in model_e_values:
                    f.write(f"    {e_val:15.6e}")
                f.write("\n")
        
        # 2. 力预测值和真值
        with open(force_values_file, "a") as f:
            # 添加当前数据集路径的标识行
            f.write(f"# {path}\n")
            for frame_idx in range(len(f0)):
                for atom_idx in range(len(f0[frame_idx])):
                    # 获取真实力值（三个方向）
                    true_fx = float(f0[frame_idx][atom_idx][0]) if np.isscalar(f0[frame_idx][atom_idx][0]) else float(np.mean(f0[frame_idx][atom_idx][0]))
                    true_fy = float(f0[frame_idx][atom_idx][1]) if np.isscalar(f0[frame_idx][atom_idx][1]) else float(np.mean(f0[frame_idx][atom_idx][1]))
                    true_fz = float(f0[frame_idx][atom_idx][2]) if np.isscalar(f0[frame_idx][atom_idx][2]) else float(np.mean(f0[frame_idx][atom_idx][2]))
                    
                    # 写入帧号、原子ID和真实力值（三个方向）
                    f.write(f"{frame_idx:12d}    {atom_idx:12d}    {true_fx:15.6e}    {true_fy:15.6e}    {true_fz:15.6e}")
                    
                    # 获取各模型预测的力值（所有方向）
                    for m in range(num_models):
                        # 计算x方向预测值
                        pred_fx = df_list[m][frame_idx][atom_idx][0] + f0[frame_idx][atom_idx][0]
                        fx_val = float(pred_fx) if np.isscalar(pred_fx) else float(np.mean(pred_fx))
                        
                        # 计算y方向预测值
                        pred_fy = df_list[m][frame_idx][atom_idx][1] + f0[frame_idx][atom_idx][1]
                        fy_val = float(pred_fy) if np.isscalar(pred_fy) else float(np.mean(pred_fy))
                        
                        # 计算z方向预测值
                        pred_fz = df_list[m][frame_idx][atom_idx][2] + f0[frame_idx][atom_idx][2]
                        fz_val = float(pred_fz) if np.isscalar(pred_fz) else float(np.mean(pred_fz))
                        
                        # 写入三个方向的预测力值
                        f.write(f"    {fx_val:15.6e}    {fy_val:15.6e}    {fz_val:15.6e}")
                    
                    f.write("\n")
        
        # 保存每个数据点的能量差值
        with open(energy_diff_file, "a") as f:
            # 添加当前数据集路径的标识行
            f.write(f"# {path}\n")
            for frame_idx in range(len(de_list[0])):
                f.write(f"{frame_idx:12d}")
                
                # 遍历所有模型
                for m in range(num_models):
                    # 检查是标量还是数组
                    if np.isscalar(de_list[m][frame_idx]):
                        e_val = float(de_list[m][frame_idx])
                    else:
                        e_val = float(np.mean(de_list[m][frame_idx]))
                    
                    # 不再除以原子数，因为差值计算已经基于转换后的能量
                    f.write(f"    {e_val:15.6e}")
                
                f.write("\n")
        
        # 保存每个数据点的力差值
        with open(force_diff_file, "a") as f:
            # 添加当前数据集路径的标识行
            f.write(f"# {path}\n")
            for frame_idx in range(len(df_list[0])):
                for atom_idx in range(len(df_list[0][frame_idx])):
                    f.write(f"{frame_idx:12d}    {atom_idx:12d}")
                    
                    # 安全地获取每个模型的力差值
                    for model_idx in range(num_models):
                        for dir_idx in range(3):  # x, y, z 三个方向
                            force_val = df_list[model_idx][frame_idx][atom_idx][dir_idx]
                            if np.isscalar(force_val):
                                f_value = float(force_val)
                            else:
                                f_value = float(np.mean(force_val))
                            f.write(f"    {f_value:15.6e}")
                    
                    f.write("\n")
        
        # 计算每个数据点的标准差
        with open(model_devi_file, "a") as f:
            # 添加当前数据集路径的标识行
            f.write(f"# {path}\n")
            for frame_idx in range(len(de_list[0])):
                # 收集当前帧的所有模型预测
                if np.isscalar(de_list[0][frame_idx]):
                    frame_e_values = [float(de_list[m][frame_idx]) for m in range(num_models)]
                else:
                    frame_e_values = [float(np.mean(de_list[m][frame_idx])) for m in range(num_models)]
                
                # 计算能量标准差（不再除以原子数）
                energy_std = np.std(frame_e_values)
                
                # 计算每个方向上的力标准差（每个原子求平均）
                fx_std_list = []
                fy_std_list = []
                fz_std_list = []
                
                for atom_idx in range(len(df_list[0][frame_idx])):
                    # 收集当前原子在所有模型上的力差值
                    atom_fx_values = []
                    atom_fy_values = []
                    atom_fz_values = []
                    
                    for model_idx in range(num_models):
                        fx_val = df_list[model_idx][frame_idx][atom_idx][0]
                        fy_val = df_list[model_idx][frame_idx][atom_idx][1]
                        fz_val = df_list[model_idx][frame_idx][atom_idx][2]
                        
                        if np.isscalar(fx_val):
                            atom_fx_values.append(float(fx_val))
                        else:
                            atom_fx_values.append(float(np.mean(fx_val)))
                            
                        if np.isscalar(fy_val):
                            atom_fy_values.append(float(fy_val))
                        else:
                            atom_fy_values.append(float(np.mean(fy_val)))
                            
                        if np.isscalar(fz_val):
                            atom_fz_values.append(float(fz_val))
                        else:
                            atom_fz_values.append(float(np.mean(fz_val)))
                    
                    # 计算每个原子在三个方向上的标准差
                    fx_std_list.append(np.std(atom_fx_values))
                    fy_std_list.append(np.std(atom_fy_values))
                    fz_std_list.append(np.std(atom_fz_values))
                
                # 计算三个方向上所有原子的平均标准差
                fx_mean_std = np.mean(fx_std_list)
                fy_mean_std = np.mean(fy_std_list)
                fz_mean_std = np.mean(fz_std_list)
                
                # 写入每个帧的标准差数据
                energy_std_float = float(energy_std)
                fx_mean_std_float = float(fx_mean_std)
                fy_mean_std_float = float(fy_mean_std)
                fz_mean_std_float = float(fz_mean_std)
                f.write(f"{frame_idx:12d}    {energy_std_float:15.6e}    {fx_mean_std_float:15.6e}    {fy_mean_std_float:15.6e}    {fz_mean_std_float:15.6e}\n")
                
        print(f"文件: {j} 处理完成")

**Comment by njzjz** [comment]
You could compare the input in `dp test` and that in your code after printing them.

https://github.com/deepmodeling/deepmd-kit/blob/70bc6d89d057d348558b87865ff2f225cae677aa/deepmd/entrypoints/test.py#L338-L348
